src/Database/Kioku/Internal/TrieIndex.hs

```diff
@@ -23,22 +24,34 @@ import            System.IO
 import            Database.Kioku.Internal.Buffer
 import            Database.Kioku.Memorizable
 
-
 trieLookup :: BS.ByteString -> TrieIndex -> [Int]
-trieLookup key = maybe [] trieRootElems . lookupSubtrie key
+trieLookup key = maybe [] (trieRootElems []) . lookupSubtrie key
 
 trieMatch :: BS.ByteString -> TrieIndex -> [Int]
 trieMatch prefix = maybe [] trieElems . lookupSubtrie prefix
 
+trieAllHitsAlong :: BS.ByteString -> TrieIndex -> [Int]
+trieAllHitsAlong = go []
+  where
+    go results "" trie = trieRootElems results trie
+    go results key trie@(TI _ _ offset) =
+      let (p, ath) = BS.splitAt 1 key
+       in case lookupSubtrie p trie of
+            Nothing -> trieRootElems results trie
+            Just subtrie@(TI _ _ subOffset) ->
+              if offset == subOffset
+                 then go results ath subtrie
+                 else go (trieRootElems results trie) ath subtrie
```

Comment:
I think extracting the `trieRootElems results trie` expression to a named variable in the `let` before the `case` would help. A good variable name would help the reader discern why we need to look up the root elements on the `Nothing` branch. (It took me several times thinking through the logic just now to convince myself)